src/web/src/app.tsx

```diff
@@ -0,0 +1,362 @@
+        <IncludeExcludeList
+          caption="Countries"
+          include={filter?.country}
+          exclude={filter?.exclude_country}
+        />
+        <IncludeExcludeList
+          caption="Hardware class"
+          include={filter?.hardware_class}
+          exclude={filter?.exclude_hardware_class}
+        />
+      </div>
+    </div>
+  );
+}
+
+function getCurrentSeedType(searchParams: URLSearchParams): SeedType {
+  return (
+    stringToSeedType(searchParams.get('seed') ?? 'PRODUCTION') ??
+    SeedType.PRODUCTION
+  );
+}
+
+export function NavItem(props: { type: SeedType }): JSX.Element {
+  const [searchParams, setSearchParams] = useSearchParams();
+  const currentSeed = getCurrentSeedType(searchParams);
+  const className =
+    (props.type === currentSeed ? 'active ' : '') + 'nav-item nav-link btn-sm';
+  const handleClick = (): void => {
+    setSearchParams((prev) => {
+      prev.set('seed', SeedType[props.type]);
+      prev.delete('search');
+      return prev;
+    });
+  };
+  return (
+    <a onClick={handleClick} className={className}>
+      {SeedType[props.type]}
+    </a>
+  );
+}
+
+export function FilterCheckbox(props: {
+  title: string;
+  htmlName: string;
+  checked: boolean;
+  toggle: () => void;
+}): JSX.Element {
+  return (
+    <div className="filter">
+      <input
+        type="checkbox"
+        id={props.htmlName}
+        checked={props.checked}
+        onChange={props.toggle}
+      />
+      <label htmlFor={props.htmlName}>{props.title}</label>
+    </div>
+  );
+}
+
+export function CurrentStudyList(props: {
+  studies: Map<SeedType, StudyListModel>;
+}): JSX.Element {
+  const [searchParams, setSearchParams] = useSearchParams();
+  const setParam = (param: string, value: string): void => {
+    setSearchParams((prev) => {
+      prev.set(param, value);
+      return prev;
+    });
+  };
+  const currentSeed = getCurrentSeedType(searchParams);
+
+  const filter = new StudyFilter();
+  filter.search = searchParams.get('search') ?? undefined;
+  filter.minPriority =
+    currentSeed === SeedType.UPSTREAM
+      ? StudyPriority.STABLE_MIN
+      : StudyPriority.NON_INTERESTING;
+  try {
+    const priorityString = searchParams.get('minPriority');
+    if (priorityString != null) {
+      filter.minPriority = parseInt(priorityString);
+    }
+  } catch {
+    /* empty */
+  }
+  filter.showEmptyGroups = searchParams.get('showEmptyGroups') === 'true';
+  const toggleShowEmptyGroups = (): void => {
+    setParam('showEmptyGroups', filter.showEmptyGroups ? 'false' : 'true');
+  };
+
+  filter.includeOutdated = searchParams.get('includeOutdated') === 'true';
+  const toggleIncludeOutdated = (): void => {
+    setParam('includeOutdated', filter.includeOutdated ? 'false' : 'true');
+  };
+
+  const studyList = props.studies
+    .get(currentSeed)
+    ?.studies(filter)
+    .map((study, i) => (
+      <StudyItem key={study.name() + i} study={study} filter={filter} />
```

Comment:
If `study.name` is guaranteed to be unique, then it's not best practice to use `i` in the key as the re-use of elements won't be as optimized. But if items never get added or removed then it's irrelevant anyway.

Reply:
Sadly, `study.name` is not unique. There is not good key for the list.
Also, we load the lists once, so we could use `index` & `seedType` as key.